module-sys/SystemManager/PowerManager.cpp

```diff
@@ -243,18 +243,37 @@ namespace sys
             return;
         }
 
+        /* Disable SysTick to prevent pending SysTick interrupt */
+        lowPowerControl->DisableSysTick();
+
+        /* Disable IRQs globally using PRIMASK and try to enter WFI mode.
+         * Do not use FreeRTOS critical section, as it will mask the interrupts
+         * required to wake the CPU up.*/
+        const auto savedPrimask   = lowPowerControl->DisableInterrupts();
         const auto timeSpentInWFI = lowPowerControl->EnterWfiModeIfAllowed();
+
+        /* Re-enable SysTick and catch up ticks if needed */
+        lowPowerControl->EnableSysTick();
         if (timeSpentInWFI > 0) {
-            /* We increase the frequency immediately after exiting WFI so that the xTaskCatchUpTicks procedure has
-             * time to execute and does not block the button press detection mechanism. */
-            SetCpuFrequency(bsp::CpuFrequencyMHz::Level_4);
-            portENTER_CRITICAL();
-            lowPowerControl->DisableSysTick();
-            xTaskCatchUpTicks(cpp_freertos::Ticks::MsToTicks(timeSpentInWFI));
-            lowPowerControl->EnableSysTick();
-            portEXIT_CRITICAL();
-            UpdateCpuFrequencyMonitor(WfiName, timeSpentInWFI);
+            /* Increase the frequency immediately after exiting WFI so that the xTaskCatchUpTicks
+             * procedure has time to execute and does not block the button press detection
+             * mechanism. A critical section is required so that any exit from any critical section
+             * inside any of the functions called here does not re-enable interrupts - unfortunately
+             * FreeRTOS critical section implementation for RT1051 completely ignores initial PRIMASK
+             * state and will re-enable interrupts even if they were disabled initially. We don't
+             * want to do that until the tick counter has been updated. */
+            cpp_freertos::CriticalSection::Enter();
+            {
+                SetCpuFrequency(bsp::CpuFrequencyMHz::Level_3);
```

Comment:
why are we changing now to `CpuFrequencyMHz::Level3` instead of `CpuFrequencyMHz::Level4`?

Reply:
After leaving critical section power management algorithms would often increase CPU frequency to 528MHz if we changed the frequency to `CpuFrequencyMHz::Level4` here. I'm not sure why that happened, but lowering the level here helped without impacting initial responsiveness and now the frequency is increased to at most 132MHz

Reply:
This may also affect power management because redrawing eInk at 132 MHz will take longer. I once did some research and it turned out that it is better to redraw at the highest frequency. We need to do the research again because power management is now significantly changed.

Reply:
@Lefucjusz take a look https://appnroll.atlassian.net/browse/MOS-117

Reply:
Before the fix I've implemented the frequency after leaving WFI would almost always switch to 132MHz, only sometimes to 528MHz, that's why I changed the CPU frequency level - to match the behavior to the current one. It's worth checking the entire power management mechanism and optimize it as it has been significantly modified indeed, but IMO it's better to do it under a separate task.

Reply:
I agree